src/tools/clippy/clippy_lints/src/methods/suspicious_map.rs

```diff
@@ -12,7 +13,7 @@ pub fn check<'tcx>(cx: &LateContext<'tcx>, expr: &hir::Expr<'_>, count_recv: &hi
     if_chain! {
         if is_trait_method(cx, count_recv, sym::Iterator);
         let closure = expr_or_init(cx, map_arg);
-        if let Some(body_id) = cx.tcx.hir().maybe_body_owned_by(closure.hir_id);
+        if let Some(body_id) = cx.tcx.hir().find(closure.hir_id).map(map::associated_body).flatten();
```

Comment:
@cjgillot I initially used `closure.hir_id.owner` but it broke the test for this function so on a whim I used `.find(closure.hir_id).map(map::associated_body).flatten()` and it worked. Why didn't `maybe_body_owned_by(LocalDefId)` worked but using the `Node` works? It looks like `maybe_body_owned_by` also uses `find` to get the `Node`.

Reply:
For closures, hir_id.owner and local_def_id(hir_id) point to different things. The former points to the enclosing function, the latter to the closure.
Using local_def_id(closure.hir_id) should work.
Is it enough to make associated_body private?

Reply:
Using `local_def_id(closure.hir_id)` does not work. Here is the error:
```
+error: internal compiler error: /home/fausto/Repo/rust/compiler/rustc_middle/src/hir/map/mod.rs:195:13: local_def_id: no entry for `HirId { owner: DefId(0:6 ~ suspicious_map[53a1]::negative), local_id: 79 }`, which has a map of `Some(Expr(Expr { hir_id: HirId { owner: DefId(0:6 ~ suspicious_map[53a1]::negative), local_id: 79 }, kind: Path(Resolved(None, Path { span: $DIR/suspicious_map.rs:27:24: 27:36 (#0), res: Def(Fn, DefId(0:10 ~ suspicious_map[53a1]::do_something)), segments: [PathSegment { ident: do_something#0, hir_id: Some(HirId { owner: DefId(0:6 ~ suspicious_map[53a1]::negative), local_id: 78 }), res: Some(Def(Fn, DefId(0:10 ~ suspicious_map[53a1]::do_something))), args: None, infer_args: true }] })), span: $DIR/suspicious_map.rs:27:24: 27:36 (#0) }))`
+
+thread 'rustc' panicked at 'Box<dyn Any>', /home/fausto/Repo/rust/compiler/rustc_errors/src/lib.rs:1392:9
+note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
+
+note: the compiler unexpectedly panicked. this is a bug.
+
+note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new
+
+note: Clippy version: clippy 0.1.64 (9b485f92000 2022-07-19)
+
+query stack during panic:
+#0 [analysis] running analysis passes on this crate
+end of query stack
+error: aborting due to 3 previous errors
```

Reply:
Aside from this, I think all tasks are done. Let me know if I missed something.

Reply:
As the code does not check whether this is a closure beforehand.  You can use `tcx.hir().opt_local_def_id` which won't ICE on failure, and just return `None`.